### ISSUE TITLE
CA1062 False positive when NotNull check not immediately before usage

### ISSUE THREAD

**Issue by gmt2001** [dotnet/roslyn-analyzers#3292]
#### Analyzer package

[Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers)

#### Package Version

v2.9.8 (Latest)

#### Diagnostic ID

[CA1062](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1062?view=vs-2019)

#### Repro steps

In the code below, CA1062 shows as being violated on the parameter `listValue` on the line `this._dictionaryList[dictionaryKey].Add(listValue.ToLowerInvariant());` on version 1, even though it has been checked, the exception thrown at the top of the method, and the object is an immutable type. Version 2 does not produce the violation.

##### Version 1
```c#
public bool AddDictionaryListElement(string dictionaryKey, string listValue)
{
    if (string.IsNullOrWhiteSpace(dictionaryKey))
    {
        throw new ArgumentNullException(nameof(dictionaryKey));
    }
    
    if (string.IsNullOrWhiteSpace(listValue))
    {
        throw new ArgumentNullException(nameof(listValue));
    }

    if (!this.DoesAnyDictionaryListHaveValue(listValue))
    {
        if (!this._dictionaryList.ContainsKey(dictionaryKey))
        {
            _ = this._dictionaryList.TryAdd(dictionaryKey, new List<string>());
        }

        this._dictionaryList[dictionaryKey].Add(listValue.ToLowerInvariant()); // CA1062

        return true;
    }

    return false;
}
```

##### Version 2

```c#
public bool AddDictionaryListElement(string dictionaryKey, string listValue)
{
    if (string.IsNullOrWhiteSpace(dictionaryKey))
    {
        throw new ArgumentNullException(nameof(dictionaryKey));
    }

    if (!this.DoesAnyDictionaryListHaveValue(listValue))
    {
        if (!this._dictionaryList.ContainsKey(dictionaryKey))
        {
            _ = this._dictionaryList.TryAdd(dictionaryKey, new List<string>());
        }
        
        if (string.IsNullOrWhiteSpace(listValue))
        {
            throw new ArgumentNullException(nameof(listValue));
        }

        this._dictionaryList[dictionaryKey].Add(listValue.ToLowerInvariant()); // No violation

        return true;
    }

    return false;
}
```

#### Expected behavior

No error due to CA1062 being violated on either version of the snippet, since string is an immutable type

#### Actual behavior

CA1062 shows as being violated on version 1 of the snippet
